File: src/FormingVectorModel/FormingVectorModel/Core/Objects/Visual/Pixel.cpp
#include "Pixel.h"
#include "../Geometric/Point.h"
#include "Color.h"

Core::Objects::Visual::Pixel::Pixel() : Pixel(nullptr, nullptr)
{
}

Core::Objects::Visual::Pixel::Pixel(IPoint* point, IColor* color) : _point(point), _color(color)
{
	if (!_point)
		_point = new Geometric::Point;

	if (!_color)
		_color = new Color;
}

Core::Objects::Visual::Pixel::Pixel(int x, int y, unsigned char red, unsigned char green, unsigned char blue) : Pixel(new Geometric::Point(x, y), new Color(red, green, blue))
{
}

Core::Objects::Visual::Pixel::Pixel(unsigned char red, unsigned char green, unsigned char blue) : Pixel(0, 0, red, green, blue)
{
}

Core::Objects::Visual::Pixel::Pixel(int x, int y, unsigned char brightness) : Pixel(new Geometric::Point(x, y), new Color(brightness))
{
}

Core::Objects::Visual::Pixel::Pixel(IPixel* other) : Pixel(other->x(), other->y(), other->red(), other->green(), other->blue())
{
}

Core::Objects::Visual::Pixel::~Pixel()
{
	if (_point)
		delete _point;

	if (_color)
		delete _color;
}

int Core::Objects::Visual::Pixel::x() const
{
	return _point->x();
}

int Core::Objects::Visual::Pixel::y() const
{
	return _point->y();
}

unsigned char Core::Objects::Visual::Pixel::red() const
{
	return _color->red();
}

void Core::Objects::Visual::Pixel::red(unsigned char value)
{
	_color->red(value);
}

unsigned char Core::Objects::Visual::Pixel::green() const
{
	return _color->green();
}

void Core::Objects::Visual::Pixel::green(unsigned char value)
{
	_color->green(value);
}

unsigned char Core::Objects::Visual::Pixel::blue() const
{
	return _color->blue();
}

void Core::Objects::Visual::Pixel::blue(unsigned char value)
{
	_color->blue(value);
}

unsigned char Core::Objects::Visual::Pixel::brightness() const
{
	return _color->brightness();
}

void Core::Objects::Visual::Pixel::brightness(unsigned char value)
{
	_color->brightness(value);
}

File: src/FormingVectorModel/FormingVectorModelLibrary/Core/Objects/Visual/Pixel.cs
using FormingVectorModelLibrary.Core.Objects.Geometric;

namespace FormingVectorModelLibrary.Core.Objects.Visual
{
    public class Pixel : IPixel
    {
        private IPoint point;
        private IColor color;

        public Pixel(int x, int y, byte red, byte green, byte blue)
        {
            point = new Point(x, y);
            color = new Color(red, green, blue);
        }

        public Pixel(IPoint point, IColor color)
        {
            this.point = point ?? new Point();
            this.color = color ?? new Color();
        }

        public int X => point.X;

        public int Y => point.Y;

        public byte Red
        {
            get { return color.Red; }
            set { color.Red = value; }
        }

        public byte Green
        {
            get { return color.Green; }
            set { color.Green = value; }
        }

        public byte Blue
        {
            get { return color.Blue; }
            set { color.Blue = value; }
        }

        public byte Brightness
        {
            get { return color.Brightness; }
            set { color.Brightness = value; }
        }
    }
}